src/sql-lexer/src/lexer.rs

```diff
@@ -75,10 +76,65 @@ impl LexerError {
     }
 }
 
+/// Newtype wrapper around [`String`] whose _byte_ length is guaranteed to be less than or equal to
+/// [`MAX_IDENTIFIER_LENGTH`].
+#[derive(Debug, Clone, PartialEq)]
+
+pub struct SmallString(String);
+
+impl SmallString {
+    /// Creates a new [`SmallString`] returning an error if `s` is more than
+    /// [`MAX_IDENTIFIER_LENGTH`] bytes long.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use mz_sql_lexer::lexer::SmallString;
+    ///
+    /// let good = SmallString::new("hello".to_string()).unwrap();
+    /// assert_eq!(good.as_str(), "hello");
+    ///
+    /// // Note: this is only 64 characters, but each character requires 4 bytes.
+    /// let too_long = "😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊😊";
+    /// let smol = SmallString::new(too_long.to_string());
+    /// assert!(smol.is_err());
+    /// ```
+    ///
+    pub fn new(s: String) -> Result<Self, String> {
+        if s.len() > MAX_IDENTIFIER_LENGTH {
+            return Err(s);
+        }
+
+        Ok(SmallString(s))
```

Comment:
What's the reason for doing this check in both the Lexer and Parser?

Reply:
The lexer is what creates `Token`s, so it's cleanest to do the check there for SQL parsing. Then I was met with a dilemma about do we move the `Ident` struct one layer lower into `mz_sql_lexer` or create this second type? 

We could probably move it one layer lower, and then re-export it from the `sql-parser` crate, but this felt a little cleaner? Not sure, what do you think?